design/pipeline-caching.md

```diff
@@ -0,0 +1,366 @@
+# Pipeline Caching
+
+*** Status: Work in Progress ***
+
+We have observed the need for Azure Pipelines to implement a caching mechanism that allows the outputs of steps within the pipeline to be skipped if a suitable cached version of the output already exists.
+
+Provided the cost of determining a cache hit and acquiring the contents of the cache is cheaper than the cost of producing the output again from scratch then the pipeline's performance will be increased.
+
+Caching can introduce many complexities into build pipelines and our goal is to strike the balance between "correct" and "good enough" where "good enough" allows us to provide performance benefits without a disproporionate amount of effort on behalf of the user.
+
+## Proposal
+
+This proposal contains a number of interelated elements. I've tried to build it up sequentially but sometimes you'll need to read further along to understand why something is the way it is.
+
+### The basics & task integration
```

Comment:
I don't think this is going to be correct in many cases. For the most part you will likely still want to run the task to normalize the cache to make sure you have the right versions as there are a lot of cases where you might choose a key like branchname and that would give you mostly the right files and then you only have to download the delta from the package source.

I also don't consider this to be a generic feature of tasks, I think having the ability to simply cache a set of files by a key and restore that is the baseline functionality and could work equally well for scenarios like build outputs as well as package cache pre-population